crypto/o_str.c

```diff
@@ -90,6 +90,69 @@ size_t OPENSSL_strlcat(char *dst, const char *src, size_t size)
     return l + OPENSSL_strlcpy(dst, src, size);
 }
 
+/**
+ * @brief Converts a string to an unsigned long integer.
+ *
+ * This function attempts to convert a string representation of a number
+ * to an unsigned long integer, given a specified base. It also provides
+ * error checking and reports whether the conversion was successful.
+ * This function is just a wrapper around the POSIX strtoul function with
+ * additional error checking.  This implies that errno for the caller is set
+ * on calls to this function.
+ *
+ * @param str The string containing the representation of the number.
+ * @param endptr A pointer to a pointer to character. If not NULL, it is set
+ *               to the character immediately following the number in the
+ *               string.
+ * @param base The base to use for the conversion, which must be between 2,
+ *             and 36 inclusive, or be the special value 0. If the base is 0,
+ *             the actual base is determined by the format of the initial
+ *             characters of the string.
+ * @param num A pointer to an unsigned long where the result of the
+ *            conversion is stored.
+ *
+ * @return 1 if the conversion was successful, 0 otherwise. Conversion is
+ *         considered unsuccessful if no digits were consumed or if an error
+ *         occurred during conversion.
+ *
+ * @note It is the caller's responsibility to check if the conversion is
+ *       correct based on the expected consumption of the string as reported
+ *       by endptr.
+ */
+int OPENSSL_strtoul(const char *str, char **endptr, int base,
+                    unsigned long *num)
+{
+    char *tmp_endptr;
+    char **internal_endptr = endptr == NULL ? &tmp_endptr : endptr;
+
+    errno = 0;
+
+    *internal_endptr = (char *)str;
+
+    if (num == NULL)
+        return 0;
+
+    if (str == NULL)
+        return 0;
+
+    /* Fail on negative input */
+    if (*str == '-')
+        return 0;
+
+    *num = strtoul(str, internal_endptr, base);
+    /*
+     * If no digits were consumed in the translation
+     * or if there was a reported error during the translation
+     * fail the operation, otherwise succeed.
+     * Note, it's the responsibility of the caller to determine if the
+     * conversion was correct based on the expected consumption of the string
+     * as reported by endptr
+     */
+    if (str == (const char *)*internal_endptr || errno != 0)
```

Comment:
```suggestion
    if (str == (const char *)*internal_endptr || errno != 0 || **internal_endptr)
```
We should also check for this. I don't see any valid use case for converting half-baked number. This will make our code more robust.

Reply:
Its always going to be the case though, a user passes a NULL endptr, internal endptr gets set to &tmp_endptr

Reply:
I mean - `I don't see any valid use case for converting half-baked number. ` 

I don't see a legitimate reason to convert a partial or malformed number.

Converting an incomplete or invalid number, such as `str = 123abc`, doesn't seem to have a practical application.

Reply:
There are valid use cases for converting strings containing something after a number. For example converting a value with magnitude. I.e., 1K, 3M, 1024G. We could actually use the input `endptr` as the discriminator - if it is set to NULL we enforce that the number ends at the end of the input string, if it is not set to NULL we do not enforce it here and leave it up to the caller.

Reply:
+1 to what @t8m says

Reply:
ACK, config files as well occasionally contain delimited series of numbers, like oids, or just value sequences:
```
1.2.3.4.5
or
12345:6789:0
```